protocol/adapters/datasync_client.go

```diff
@@ -0,0 +1,223 @@
+type DataSyncClient struct {
+	sync *mvds.Node
+	t    *DataSyncWhisperTransport
+}
+
+func NewDataSyncClient(sync *mvds.Node, t *DataSyncWhisperTransport) *DataSyncClient {
+	go sync.Run()
+
+	return &DataSyncClient{
+		sync: sync,
+		t:    t,
+	}
+}
+
+// Subscribe subscribes to a public chat using the Whisper service.
+func (c *DataSyncClient) Subscribe(ctx context.Context, messages chan<- *protocol.Message, options protocol.SubscribeOptions) (*protocol.Subscription, error) {
+	return c.t.subscribe(messages, options)
+}
+
+// Send appends a message to the data sync node for later sending.
+func (c *DataSyncClient) Send(ctx context.Context, data []byte, options protocol.SendOptions) ([]byte, error) {
+	if err := options.Validate(); err != nil {
+		return nil, err
+	}
+
+	topic, err := topic(options)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := c.sync.AppendMessage(toGroupId(topic), data)
+	if err != nil {
+		return nil, err
+	}
+
+	return id[:], nil
+}
+
+func (*DataSyncClient) Request(ctx context.Context, params protocol.RequestOptions) error {
+	return nil
+}
+
+type DataSyncWhisperTransport struct {
+	shh         *whisper.Whisper
+	keysManager *whisperServiceKeysManager
+
+	packets chan mvds.Packet
+}
+
+func NewDataSyncWhisperTransport(shh *whisper.Whisper, privateKey *ecdsa.PrivateKey) *DataSyncWhisperTransport {
+	return &DataSyncWhisperTransport{
+		shh: shh,
+		keysManager: &whisperServiceKeysManager{
+			shh:               shh,
+			privateKey:        privateKey,
+			passToSymKeyCache: make(map[string]string),
+		},
+		packets: make(chan mvds.Packet),
+	}
+}
+
+func (t *DataSyncWhisperTransport) Watch() mvds.Packet {
+	return <-t.packets
+}
+
+// Send sends a new message using the Whisper service.
+func (t *DataSyncWhisperTransport) Send(group mvds.GroupID, _ mvds.PeerId, peer mvds.PeerId, payload mvds.Payload) error {
+	data, err := proto.Marshal(&payload)
+
+	newMessage, err := newNewMessage(t.keysManager, data)
+	if err != nil {
+		return err
+	}
+
+	newMessage.Topic = toTopicType(group)
+
+	// @todo set SymKeyID or PublicKey depending on chat type
+
+	// we are only assuming private chats
+	k := ecdsa.PublicKey(peer)
+	newMessage.PublicKey = crypto.FromECDSAPub(&k)
+
+	_, err = whisper.NewPublicWhisperAPI(t.shh).Post(context.Background(), newMessage.ToWhisper())
+	return err
+}
+
+func (t *DataSyncWhisperTransport) subscribe(in chan<- *protocol.Message, options protocol.SubscribeOptions) (*protocol.Subscription, error) {
+	if err := options.Validate(); err != nil {
+		return nil, err
+	}
+
+	filter := newFilter(t.keysManager)
+	if err := updateFilterFromSubscribeOptions(filter, options); err != nil {
+		return nil, err
+	}
+
+	filterID, err := t.shh.Subscribe(filter.ToWhisper())
+	if err != nil {
+		return nil, err
+	}
+
+	subWhisper := newWhisperSubscription(t.shh, filterID)
+	sub := protocol.NewSubscription()
+
+	go func() {
+		defer subWhisper.Unsubscribe() // nolint: errcheck
+
+		tick := time.NewTicker(time.Second)
```

Comment:
I would say that syncing once a second might be slow... 0.3 seconds is a good threshold.

Reply:
@mandrigin I left this at 1 second because it seems to be what the other adapters use too.